### Import Necessary Libraries

In [9]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
sns.set(color_codes=True)
import glob

### Read in the Dataset

In [10]:
# Point to data files location
path = "./scraper/data/6_23_2025/"
all_files = glob.glob(path + "*.json")

# Read all files into single dataframe
df_list = []
for file in all_files:
    df = pd.read_json(file)
    df_list.append(df)

# ignore_index resets index of combined DF to prevent index duplicates
raw_df = pd.concat(df_list, ignore_index=True) 

# Display entire data frame
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)

# Examine and excerpt of data
raw_df.head(5)

stockNumber                vin  year        make              model  \
0     27461478  5NPEG4JA8NH134099  2022     Hyundai             Sonata   
1     26070459  JTDEPMAE5NJ210260  2022      Toyota            Corolla   
2     27175122  3GCUDFE80RG128481  2024   Chevrolet     Silverado 1500   
3     27072297  1V2WE2CA5PC229115  2023  Volkswagen  Atlas Cross Sport   
4     27386026  58AGZ1B12MU110647  2021       Lexus             ES 350   

               body           trim  basePrice  originalPrice  hasPriceDrop  \
0          4D Sedan             SE      19998            NaN         False   
1          4D Sedan             LE      20998            NaN         False   
2       4D Crew Cab  LT Trail Boss      51998            NaN         False   
3  4D Sport Utility      SE w/Tech      32998            NaN         False   
4          4D Sedan        F-Sport      35998            NaN         False   

   mileage  storeId storeName storeCity    state stateAbbreviation  distance  \
0    56782     7298   Roswell   Roswell  Georgia                GA    2108.0   
1    33730     7298   Roswell   Roswell  Georgia                GA    2108.0   
2    58240     7298   Roswell   Roswell  Georgia                GA    2108.0   
3    15895     7298   Roswell   Roswell  Georgia                GA    2108.0   
4    26900     7298   Roswell   Roswell  Georgia                GA    2108.0   

   averageRating  numberOfReviews  repairPalData  isNewArrival  \
0       4.200000                5            NaN         False   
1       4.500000               22            NaN         False   
2       5.000000                1            NaN         False   
3       5.000000                2            NaN         False   
4       4.333333                6            NaN         False   

   isTransferable                                           features  \
0            True  [Rear Defroster, Overhead Airbags, Auxiliary A...   
1            True  [Power Windows, Power Locks, Air Conditioning,...   
2            True  [Power Locks, Auxiliary Audio Input, AM/FM Ste...   
3            True  [Alloy Wheels, 20 Inch Plus Wheels, Power Lock...   
4            True  [Power Windows, Traction Control, ABS Brakes, ...   

                                 highlightedFeatures  \
0  Rear View Camera,Auxiliary Audio Input,Alloy W...   
1  Rear View Camera,Auxiliary Audio Input,Overhea...   
2  Z71 Package,4WD/AWD,Leather Seats,BOSE Sound S...   
3  Technology Package,Leatherette Seats,Satellite...   
4  Leatherette Seats,Pioneer Sound System,Parking...   

                                          highlights  \
0  [singleOwner, warranty, fuelEfficient, advance...   
1        [warranty, fuelEfficient, advancedFeatures]   
2      [singleOwner, advancedFeatures, premiumAudio]   
3  [singleOwner, lowMiles, warranty, advancedFeat...   
4  [singleOwner, lowMilesPerYear, warranty, advan...   

       lastMadeSaleableDate  transferFee            transferTags  \
0  2025-06-12T20:56:13.565Z       1999.0  [GeneratedByCps, Paid]   
1  2025-06-12T20:53:33.311Z       1999.0  [GeneratedByCps, Paid]   
2  2025-06-20T21:39:32.178Z       1999.0  [GeneratedByCps, Paid]   
3  2025-05-28T20:50:22.353Z       1999.0  [GeneratedByCps, Paid]   
4   2025-06-13T20:59:37.45Z       1999.0  [GeneratedByCps, Paid]   

     transferText   transferType  minEstimatedTransferDurationInDays  \
0  $1999 Shipping  shipping-from                                 9.0   
1  $1999 Shipping  shipping-from                                 9.0   
2  $1999 Shipping  shipping-from                                 9.0   
3  $1999 Shipping  shipping-from                                 9.0   
4  $1999 Shipping  shipping-from                                 9.0   

   maxEstimatedTransferDurationInDays  transferTimesUnavailable exteriorColor  \
0                                22.0                     False          Gray   
1                                22.0                     False          Blue   
2                  

### Check Raw Dataframe Shape

In [11]:
print(raw_df.shape)

(87118, 69)


Here we can see that the initial dataframe consists of **87,118** rows and **69** columns

Since the purpose of this project is to build a predictive model capable of accurately estimating a car's value based on a particular set of features, it is important to determine which feature will serve as the **response variable** (aka dependent variable) and which set of features will serve as the **predictors** (aka independent variables)

### Checking & Converting Feature DataTypes

In [12]:
# Check data types of all features
pd.set_option('display.max_rows', 69)
raw_df.dtypes

stockNumber                             int64
vin                                    object
year                                    int64
make                                   object
model                                  object
body                                   object
trim                                   object
basePrice                               int64
originalPrice                         float64
hasPriceDrop                             bool
mileage                                 int64
storeId                                 int64
storeName                              object
storeCity                              object
state                                  object
stateAbbreviation                      object
distance                              float64
averageRating                         float64
numberOfReviews                         int64
repairPalData                         float64
isNewArrival                             bool
isTransferable                    

Based on the features listed in the raw dataframe above, it is plausible that the following features are unlikely to have significant predictive power towards the task at hand. These irrelevant columns are as follows:

- stockNumber
- originalPrice
- hasPriceDrop
- minEstimatedTransferDurationInDays
- transferFee
- maxEstimatedTransferDurationInDays
- isNewArrival
- isTransferable
- transferText
- transferTimesUnavailable
- numberOfFavorites
- isSaleable
- isComingSoon
- isReserved
- isAvailableToTransact
- store
- isSavedCar
- heroImageUrl
- isEVTaxCreditEligible
- lastMadeSaleableDate
- transferTags
- transferType
- heroThumbnailImageUrl
- normalizedInteriorColor
- normalizedExteriorColor